streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -148,13 +199,33 @@ protected <R> QueryResult<R> runRangeQuery(final Query<R> query,
             throw new UnsupportedOperationException("Versioned stores do not support RangeQuery queries at this time.");
         }
 
-        @Override
-        protected <R> QueryResult<R> runKeyQuery(final Query<R> query,
-                                                 final PositionBound positionBound,
-                                                 final QueryConfig config) {
-            // throw exception for now to reserve the ability to implement this in the future
-            // without clashing with users' custom implementations in the meantime
-            throw new UnsupportedOperationException("Versioned stores do not support KeyQuery queries at this time.");
+        @SuppressWarnings("unchecked")
+        protected <R> QueryResult<R> runVersionedKeyQuery(final Query<R> query,
+            final PositionBound positionBound,
+            final QueryConfig config) {
+            if (query instanceof VersionedKeyQuery) {
+                final QueryResult<R> result;
+                final VersionedKeyQuery<K, V> typedKeyQuery = (VersionedKeyQuery<K, V>) query;
+                VersionedKeyQuery<Bytes, byte[]> rawKeyQuery;
+                rawKeyQuery = VersionedKeyQuery.withKey(keyBytes(typedKeyQuery.key()));
+                if (typedKeyQuery.asOfTimestamp().isPresent()) {
+                    rawKeyQuery = rawKeyQuery.asOf(typedKeyQuery.asOfTimestamp().get());
+                }
+                final QueryResult<VersionedRecord<byte[]>> rawResult =
+                    wrapped().query(rawKeyQuery, positionBound, config);
+                if (rawResult.isSuccess()) {
+                    final VersionedRecord<V> versionedRecord = VersionedStoreQueryUtils.deserializeVersionedRecord(plainValueSerdes, rawResult.getResult());
+                    final QueryResult<VersionedRecord<V>> typedQueryResult =
+                        InternalQueryResultUtil.copyAndSubstituteDeserializedResult(rawResult, versionedRecord);
+                    result = (QueryResult<R>) typedQueryResult;
+                } else {
+                    // the generic type doesn't matter, since failed queries have no result set.
+                    result = (QueryResult<R>) rawResult;
+                }
+                return result;
+            }
+            // reserved for other IQv2 query types (KIP-968, KIP-969)
```

Comment:
Actually wondering if this is the right approach? If we add new query types, we would add new query handler and new methods, right? So we don't need the `if (query instanceof VersionedKeyQuery)` to begin with, because we would call `runVersionedKeyQuery` only if the type was `VerionedKeyQuery` to begin with (then query handler ensure this part)?